### PR TITLE
Support fused batchnorm with any ndims and axis

### PR DESCRIPTION
**Pull request by benbarsdell**
- Modifies the BatchNormalizationBase layer to support fused=True with any number of dimensions and any (single) axis by adding reshape operations around the calls to nn.fused_batch_norm (which itself only supports NCHW and NHWC tensors).
- Fixes tests to account for the broader support and adds a new test for a 3D convnet.
- Use of fused=True is particularly important for mixed-precision training.

cc @reedwm @nluehr

### CODE REVIEW

File: tensorflow/python/keras/layers/normalization.py [hunk 1]
```diff
@@ -281,6 +273,51 @@ def _support_zero_size_input(self):
...
+        # Transform [..., C] to [..., 1..., C] (ndims=4).
+        shape = shape[:-1] + new_dims + shape[-1:]
+      else:
+        # Transform [N, C, ...] to [N, C, ..., 1...] (ndims=4).
+        shape += new_dims
+    elif ndims > 4:
+      # Merge excess spatial dims into the second spatial dim.
+      # Transform [N, C, H, W, ...] to [N, C, H, product(W, ...)].
+      # Or        [N, H, W, ..., C] to [N, H, product(W, ...), C].
+      merge_dim = 2 if is_channels_last else 3
```
Comment (reedwm) [inline tensorflow/python/keras/layers/normalization.py:314]: Similar to above, maybe do:

```python
product = 1
for elem in shape[merge_dim:merg_dim + ndims - 4]:
  product *= elem
shape[merge_dim:merg_dim + ndims - 4] = [product]
```

File: tensorflow/python/keras/layers/normalization.py [hunk 2]
```diff
@@ -315,18 +352,22 @@ def build(self, input_shape):
...
-          raise ValueError('Batch normalization layers with fused=True only '
-                           'support 4D input tensors.')
-      else:
+      if len(self.axis) == 1 and (self.axis[0] not in (1, ndims - 1) or
+                                  ndims != 4):
+        # The fused implementation only supports NCHW or NHWC, so we will
+        # reshape the input/output tensor to/from an equivalent 4D shape.
+        fused_shape, fused_axis = self._get_shape_and_axis_for_fused(
+            input_shape.dims, self.axis[0])
+        fused_shape = tensor_shape.TensorShape(fused_shape)
```
Comment (reedwm) [inline tensorflow/python/keras/layers/normalization.py:364]: why convert this to a TensorShape then immediately convert back to a list?

File: tensorflow/python/keras/layers/normalization.py [hunk 3]
```diff
@@ -315,18 +352,22 @@ def build(self, input_shape):
...
+        self._input_fused_shape = [-1] + fused_shape.as_list()[1:]
```
Comment (reedwm) [inline tensorflow/python/keras/layers/normalization.py:365]: The issue with computing this in `build()` is that the input shape might change between calls to the layer. The channel dimension must stay the same since that affects the size of the weights, but other dimensions can differ.

Instead, compute this in `call`.

File: tensorflow/python/keras/layers/normalization_test.py [hunk 4]
```diff
@@ -138,6 +138,27 @@ def test_batchnorm_convnet_channel_last(self):
     np.testing.assert_allclose(np.mean(out, axis=(0, 1, 2)), 0.0, atol=1e-1)
     np.testing.assert_allclose(np.std(out, axis=(0, 1, 2)), 1.0, atol=1e-1)
 
+  @keras_parameterized.run_all_keras_modes
+  def test_batchnorm_convnet_channel_last_3d_fused(self):
```
Comment (reedwm) [inline tensorflow/python/keras/layers/normalization_test.py:142]: I would also add a test with simple hand-written inputs. No need to test gradients. I'm worried this won't catch errors in `_get_shape_and_axis_for_fused`. Note gamma and beta default to ones and zeros, so you can effectively ignore them in tests that don't take gradients.

**Review by reedwm — CHANGES_REQUESTED** [review]
Thanks for the PR!

File: tensorflow/python/keras/layers/normalization.py [hunk 5]
```diff
@@ -315,18 +352,22 @@ def build(self, input_shape):
         raise ValueError('When using virtual_batch_size, adjustment cannot '
                          'be specified')
 
+    fused_axis = self.axis
+    self._input_fused_shape = None
     if self.fused in (None, True):
```
[the remark this answers is not in the document]
Comment (benbarsdell) [reply tensorflow/python/keras/layers/normalization.py]: Done.

File: tensorflow/python/keras/layers/normalization.py [hunk 6]
```diff
@@ -315,18 +352,22 @@ def build(self, input_shape):
...
+                                  ndims != 4):
+        # The fused implementation only supports NCHW or NHWC, so we will
+        # reshape the input/output tensor to/from an equivalent 4D shape.
+        fused_shape, fused_axis = self._get_shape_and_axis_for_fused(
+            input_shape.dims, self.axis[0])
+        fused_shape = tensor_shape.TensorShape(fused_shape)
+        self._input_fused_shape = [-1] + fused_shape.as_list()[1:]
+        fused_axis = [fused_axis]
+
+      if not self._USE_V2_BEHAVIOR:
```
[the remark this answers is not in the document]
Comment (benbarsdell) [reply tensorflow/python/keras/layers/normalization.py]: Done.

File: tensorflow/python/keras/layers/normalization.py [hunk 7]
```diff
@@ -281,6 +273,51 @@ def _support_zero_size_input(self):
         distribution_strategy_context.get_strategy().extended,
         'experimental_enable_get_next_as_optional', False)
 
+  def _get_shape_and_axis_for_fused(self, nd_shape, nd_axis):
+    '''Returns a 4D shape and axis (1 or 3) to which nd_shape and nd_axis can
```
[the remark this answers is not in the document]
Comment (benbarsdell) [reply tensorflow/python/keras/layers/normalization.py]: Done.

File: tensorflow/python/keras/layers/normalization.py [hunk 8]
```diff
@@ -281,6 +273,51 @@ def _support_zero_size_input(self):
...
+    be changed without changing the result of the batch normalization operation.
+    '''
+    assert(isinstance(nd_axis, int))
```
[the remark this answers is not in the document]
Comment (benbarsdell) [reply tensorflow/python/keras/layers/normalization.py:291]: `isinstance(axis, int)` is already used in several places in this class, including in an error check at the beginning of the `__init__` function.

File: tensorflow/python/keras/layers/normalization.py [hunk 9]
```diff
@@ -281,6 +273,51 @@ def _support_zero_size_input(self):
...
+    if axis not in (1, ndims - 1):
+      # Move axis to dim 1.
+      if axis == 0:
+        # Transform [C, ...] to [1, C, ...].
+        shape.insert(0, 1)
+        ndims += 1
+      else:
+        # Merge excess pre-axis dims into first dim.
+        # Transform [N, ..., C, ...] to [product(N, ...), C, ...].
+        for dim in range(axis - 1, 0, -1):
```
[the remark this answers is not in the document]
Comment (benbarsdell) [reply tensorflow/python/keras/layers/normalization.py]: Done.

File: tensorflow/python/keras/layers/normalization.py (hunk 1 above)
Comment (benbarsdell) [reply tensorflow/python/keras/layers/normalization.py:326]: Done.

File: tensorflow/python/keras/layers/normalization.py (hunk 2 above)
Comment (benbarsdell) [reply tensorflow/python/keras/layers/normalization.py]: Done. (Twas a remnant of a fight I had initially with TensorShape vs. list, None vs. -1 etc.).

File: tensorflow/python/keras/layers/normalization.py (hunk 3 above)
Comment (benbarsdell) [reply tensorflow/python/keras/layers/normalization.py]: Done.

File: tensorflow/python/keras/layers/normalization_test.py (hunk 4 above)
Comment (benbarsdell) [reply tensorflow/python/keras/layers/normalization_test.py:142]: Done.

**Review by reedwm — CHANGES_REQUESTED** [review]

**Comment by gbaned** [comment]
@benbarsdell Can you please check @reedwm's comments and keep us posted. Thanks!

File: tensorflow/python/keras/layers/normalization.py [hunk 10]
```diff
@@ -315,33 +365,9 @@ def build(self, input_shape):
...
-
-    if self.fused:
-      if self.axis == [1]:
-        self._data_format = 'NCHW'
-      elif self.axis == [3]:
-        self._data_format = 'NHWC'
-      else:
-        raise ValueError('Unsupported axis, fused batch norm only supports '
-                         'axis == [1] or axis == [3]')
+    self._input_fused_shape = None
```
[the remark this answers is not in the document]
Comment (benbarsdell) [reply tensorflow/python/keras/layers/normalization.py]: Fixed.

File: tensorflow/python/keras/layers/normalization.py [hunk 11]
```diff
@@ -499,6 +525,27 @@ def _fused_batch_norm(self, inputs, training):
     beta = self.beta if self.center else self._beta_const
     gamma = self.gamma if self.scale else self._gamma_const
 
+    original_shape = None
+    fused_axis = self.axis[0]
+    input_shape = inputs.shape.as_list()
```
[the remark this answers is not in the document]
Comment (benbarsdell) [reply tensorflow/python/keras/layers/normalization.py]: Done.

File: tensorflow/python/keras/layers/normalization_test.py [hunk 12]
```diff
@@ -153,11 +153,48 @@ def test_batchnorm_convnet_channel_last_3d_fused(self):
...
+        optimizer=gradient_descent.GradientDescentOptimizer(0.01))
+
+    # Sequential values ensure the result is axis-dependent.
+    x = np.arange(5 * 5 * 5 * 5 * 3).reshape([5, 5, 5, 5, 3])
+    x = x.astype(np.float32)
+    model.fit(x, x, epochs=1000, verbose=0)
+    moving_mean = keras.backend.eval(norm.moving_mean)
+    moving_variance = keras.backend.eval(norm.moving_variance)
+    np.testing.assert_allclose(
+        moving_mean, np.array([936., 937., 938.]), rtol=1e-5)
```
[the remark this answers is not in the document]
Comment (benbarsdell) [reply tensorflow/python/keras/layers/normalization_test.py]: I've changed it to compute the target values on the fly so that it's clear where they come from.

**Comment by benbarsdell** [comment]
I managed to root-cause the `layer_correctness_test` failure. There appears to be a bug [here](https://github.com/tensorflow/tensorflow/blob/9c7ddffd970e8e72585c66c4a2c7cf8bb42ed608/tensorflow/core/kernels/redux_functor.h#L60) (also [here](https://github.com/tensorflow/tensorflow/blob/9c7ddffd970e8e72585c66c4a2c7cf8bb42ed608/tensorflow/core/kernels/redux_functor.h#L229)) where the input is interpreted as `OutputT` instead of `InputT`.
I'm not sure how to add a tensor typecast there, but just removing that piece of code does fix the problem.

Let me know how you want to deal with it.

File: tensorflow/python/keras/layers/normalization.py [hunk 13]
```diff
@@ -281,53 +281,55 @@ def _get_shape_and_axis_for_fused(self, nd_shape, nd_axis):
...
+        shape = array_ops.concat([constant_op.constant([1]), shape], axis=0)
         ndims += 1
       else:
         # Merge excess pre-axis dims into first dim.
         # Transform [N, ..., C, ...] to [product(N, ...), C, ...].
-        product = 1
-        for dim in shape[:axis]:
-          product *= dim
-        shape[:axis] = [product]
+        product = math_ops.reduce_prod(shape[:axis])
```
Comment (reedwm) [inline tensorflow/python/keras/layers/normalization.py:305]: If you pass `keepdims=True` here, I think you don't need to call `array_ops.reshape` on the line below. This makes it slightly simpler.

And same for the place below you call `math_ops.reduce_prod`

File: tensorflow/python/keras/layers/normalization.py [hunk 14]
```diff
@@ -527,15 +528,18 @@ def _fused_batch_norm(self, inputs, training):
...
+    ndims = len(inputs.shape)
     if self.axis[0] not in (1, ndims - 1) or ndims != 4:
       # The fused implementation only supports NCHW or NHWC, so we reshape the
       # input/output tensor to/from an equivalent 4D shape.
       fused_shape, fused_axis = self._get_shape_and_axis_for_fused(input_shape,
                                                                    self.axis[0])
-      original_shape = [-1] + input_shape[1:]
-      inputs = array_ops.reshape(inputs, [-1] + fused_shape[1:])
+      original_shape = array_ops.concat(
+          [constant_op.constant([-1]), input_shape[1:]], axis=0)
```
Comment (reedwm) [inline tensorflow/python/keras/layers/normalization.py:539]: Can you pass `[-1]` here instead of `constant_op.constant([-1])`? The former is slightly simpler.

File: tensorflow/python/keras/layers/normalization.py [hunk 15]
```diff
@@ -527,15 +528,18 @@ def _fused_batch_norm(self, inputs, training):
...
+      fused_shape = array_ops.concat(
+          [constant_op.constant([-1]), fused_shape[1:]], axis=0)
```
Comment (reedwm) [inline tensorflow/python/keras/layers/normalization.py:541]: Why do you need to replace the first element of `fused_shape` with -1?

File: tensorflow/python/keras/layers/normalization_test.py [hunk 16]
```diff
@@ -213,7 +277,7 @@ def call(self, x, training):
       model = MyModel()
 
       for _ in range(10):
-        x = constant_op.constant(0.5, shape=[1, 1])
+        x = constant_op.constant(0.5, shape=[2, 1])
```
Comment (reedwm) [inline tensorflow/python/keras/layers/normalization_test.py:280]: Why this change?

File: tensorflow/python/keras/layers/normalization_test.py [hunk 17]
```diff
@@ -138,6 +138,70 @@ def test_batchnorm_convnet_channel_last(self):
...
+    model.fit(x, x, epochs=4, verbose=0)
+    out = model.predict(x)
+    out -= np.reshape(keras.backend.eval(norm.beta), (1, 1, 1, 1, 3))
+    out /= np.reshape(keras.backend.eval(norm.gamma), (1, 1, 1, 1, 3))
+
+    np.testing.assert_allclose(np.mean(out, axis=(0, 1, 2, 3)), 0.0, atol=1e-1)
+    np.testing.assert_allclose(np.std(out, axis=(0, 1, 2, 3)), 1.0, atol=1e-1)
+
+  @keras_parameterized.run_all_keras_modes
+  def test_batchnorm_convnet_channel_last_3d_fused_correctness(self):
```
Comment (reedwm) [inline tensorflow/python/keras/layers/normalization_test.py:163]: I think there are no non-4D fused tests where the axis isn't the last element. Either change this test or the previous to have the axis be not the last element, or add a new test.

**Review by reedwm — CHANGES_REQUESTED** [review]

**Comment by reedwm** [comment]
> I managed to root-cause the `layer_correctness_test` failure. There appears to be a bug [here](https://github.com/tensorflow/tensorflow/blob/9c7ddffd970e8e72585c66c4a2c7cf8bb42ed608/tensorflow/core/kernels/redux_functor.h#L60) (also [here](https://github.com/tensorflow/tensorflow/blob/9c7ddffd970e8e72585c66c4a2c7cf8bb42ed608/tensorflow/core/kernels/redux_functor.h#L229)) where the input is interpreted as `OutputT` instead of `InputT`.
> I'm not sure how to add a tensor typecast there, but just removing that piece of code does fix the problem.
> 
> Let me know how you want to deal with it.

Can you create a new PR to fix it? [This line](https://github.com/tensorflow/tensorflow/blob/9c7ddffd970e8e72585c66c4a2c7cf8bb42ed608/tensorflow/core/kernels/redux_functor.h#L195) has an example of how to cast. If this is trickier than I'm imagining and you don't want to deal with this, I can also get someone internally to fix it.

**Comment by benbarsdell** [comment]
> > I managed to root-cause the `layer_correctness_test` failure. There appears to be a bug [here](https://github.com/tensorflow/tensorflow/blob/9c7ddffd970e8e72585c66c4a2c7cf8bb42ed608/tensorflow/core/kernels/redux_functor.h#L60) (also [here](https://github.com/tensorflow/tensorflow/blob/9c7ddffd970e8e72585c66c4a2c7cf8bb42ed608/tensorflow/core/kernels/redux_functor.h#L229)) where the input is interpreted as `OutputT` instead of `InputT`.
> > I'm not sure how to add a tensor typecast there, but just removing that piece of code does fix the problem.
> > Let me know how you want to deal with it.
> 
> Can you create a new PR to fix it? [This line](https://github.com/tensorflow/tensorflow/blob/9c7ddffd970e8e72585c66c4a2c7cf8bb42ed608/tensorflow/core/kernels/redux_functor.h#L195) has an example of how to cast. If this is trickier than I'm imagining and you don't want to deal with this, I can also get someone internally to fix it.

Thanks, I tried that but unfortunately only Eigen tensors have a `cast` method; the TF Tensor type does not. At this stage it's probably easier for someone internal to fix it, unless there's some other very easy fix that I haven't identified.

File: tensorflow/python/keras/layers/normalization_test.py (hunk 16 above)
Comment (benbarsdell) [reply tensorflow/python/keras/layers/normalization_test.py:280]: The reason is... complicated :)

First, Bessel's correction (`N/(N-1)`) is used for the fused implementation but _not_ for the non-fused. There is a discussion of this [here](https://github.com/keras-team/keras/pull/9028).

Then, for the fused case, the [CPU implementation avoids returning NaN when N < 2](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/fused_batch_norm_op.cc#L176-L180), while the GPU implementation (CUDNN) explicitly returns NaN in these cases.

So when the test calls batchnorm with N = 1 and it (with this PR) calls the fused implementation, it returns NaN (on the GPU) and the test fails.
Changing the test size to N = 2 avoids this problem.

This probably isn't ever observed in the real world because batchnorm doesn't work when N is small anyway.

Let me know if you'd like me to take another action here.

File: tensorflow/python/keras/layers/normalization.py (hunk 13 above)
Comment (benbarsdell) [reply tensorflow/python/keras/layers/normalization.py]: Done.

File: tensorflow/python/keras/layers/normalization.py (hunk 14 above)
Comment (benbarsdell) [reply tensorflow/python/keras/layers/normalization.py]: Removed.

File: tensorflow/python/keras/layers/normalization.py (hunk 15 above)
Comment (benbarsdell) [reply tensorflow/python/keras/layers/normalization.py]: Good point, I've removed these operations.

File: tensorflow/python/keras/layers/normalization_test.py (hunk 17 above)
Comment (benbarsdell) [reply tensorflow/python/keras/layers/normalization_test.py:184]: I've added channels-first versions of the two tests.

**Review by reedwm — COMMENTED** [review]
This change looks good, but I am worried about using Bessel's correction in more cases now. It's possible (but unlikely) we may break someone who currently relies on Bessel's correction not being applied.

I think the best solution is to change this layer to use Bessel's correction in the unfused case before submitting this change. Then we can extensively test the Bessel's correction change before submitting, potentially rolling it back if it breaks someone. Once submitted, this change is a lot safer since it doesn't affect correctness, only performance. I can work on using Bessel's correction in the unfused case since I can run all Google tests before submitting it.

Do you know why Bessel's correction for cuDNN is only used for the moving variance, not the internal variance the input is divided by? We copy this behavior in the CPU version of FusedBatchNorm as well.

**Comment by gbaned** [comment]
@benbarsdell Can you please check @reedwm's comments and keep us posted. Thanks!

**Comment by benbarsdell** [comment]
Regarding Bessel's correction at inference time but not training time, it seems that the original batchnorm paper indicated doing it this way; there is some discussion here (evidently it is the same way in PyTorch):
https://github.com/pytorch/pytorch/issues/1410
https://stats.stackexchange.com/questions/311074/batch-normalization-variance-calculation

It sounds like you are going to take care of the remaining items internally @reedwm. If I've misinterpreted that or you need me to do anything else just let me know.

**Comment by reedwm** [comment]
Yes I will take care of using Bessel's correction for the nonfused implementation internally. Then we can merge this PR.

**Comment by gbaned** [comment]
@reedwm Any update on this PR? Please. Thanks!

**Comment by gbaned** [comment]
@reedwm Any update on this PR? Please. Thanks!

**Comment by gbaned** [comment]
@benbarsdell  Can you please resolve conflicts? Thanks!

File: tensorflow/python/keras/layers/normalization_test.py (hunk 16 above)
Comment (rmlarsen) [reply tensorflow/python/keras/layers/normalization_test.py:344]: If we know that N = 1, we should probably not allow calling the fused version on GPU. I'd prefer an error message over silently producing NaNs.

Comment (benbarsdell) [reply tensorflow/python/keras/layers/normalization_test.py:358]: Unfortunately this can't easily be done in the Python layer because it only applies in training mode. Adding a check to build() causes a bunch of tests to fail (ones that don't use training mode). It would have to be an in-graph check, which I don't think there's much precedent for(?).

Comment (rmlarsen) [reply tensorflow/python/keras/layers/normalization_test.py:358]: You can add a tf.assert in graph check here. This is not unprecedented and no different from adding a check in the kernel itself.

Comment (benbarsdell) [reply tensorflow/python/keras/layers/normalization_test.py:358]: I've added an in-graph check using Assert.

Comment (benbarsdell) [reply tensorflow/python/keras/layers/normalization_test.py:358]: It caused a bunch of other tests to fail. I'll have to look into them.

**Review by rmlarsen — APPROVED** [review]

**Review by rmlarsen — CHANGES_REQUESTED** [review]
Reed wanted to discuss a few design issues before approving.

**Comment by gbaned** [comment]
@benbarsdell Can you please resolve conflicts? Thanks!

File: tensorflow/python/keras/layers/normalization.py [hunk 18]
```diff
@@ -499,6 +524,37 @@ def _fused_batch_norm(self, inputs, training):
...
     gamma = self.gamma if self.scale else self._gamma_const
 
+    original_shape = None
+    fused_axis = self.axis[0]
+    input_shape = array_ops.shape(inputs)
+
+    # The use of Bessel's correction in training mode imposes the requirement
+    # that the number of elements in the reduced dimensions is > 1.
+    check = control_flow_ops.Assert(
+          training == False or math_ops.reduce_prod(input_shape) > self.depth,
```
Comment (reedwm) [inline tensorflow/python/keras/layers/normalization.py:534]: I think training can also be a tensor. The assert should be added to [this function](https://github.com/tensorflow/tensorflow/blob/807f17221d628b9fc722df5d99562f5ef126842c/tensorflow/python/keras/layers/normalization.py#L552) I think

Comment (reedwm) [reply tensorflow/python/keras/layers/normalization.py:534]: Also probably better to also raise an error immediately if the shape and training boolean are statically known

**Comment by benbarsdell** [comment]
It turns out that the reshapes added by this PR prevent the grappler layout optimizer from optimizing out all the NCHW<->NHWC transposes.

For this reason I think we will have to abandon this PR and use https://github.com/tensorflow/tensorflow/pull/42970 instead.
